Speed up `push` by using fast SQL-based `Branch.before` calculation

pchiusano opened the pull request:
This PR does a few things to close #1958 

* `push` no longer does a merge with the remote namespace - this isn't necessary since the thing you are pushing must be a fast forward or the `push` is rejected. So either it will be rejected or it will succeed with the result equaling the thing you're pushing. Basically, this is the observation that when `before a b`, `merge a b == b`. Just doing a `before` check rather than a full merge is a big deal since a cross-database merge isn't something we can do efficiently yet. 
* The `Branch.before` check that we do before `push` (to make sure that what you're pushing is a fast forward) is now done efficiently within the codebase.
* The two new tests demonstrate that ucm can [still] distinguish between a push to a codebase that is `before` us, vs. one that is not.

__Todo:__

- [x] During push, make sure to sync to staging codebase before calling before on staging codebase
- [x] @aryairani to audit and/or possibly improve the `SqlCodebase.beforeImpl` implementation and replace it with some SQL wizardry if the current LCA-based implementation is no good.
- [x] update `before` query to short-circuit per https://sqlite.org/forum/forumpost/ed2d66f197?t=h
- [x] verify that we have tests that will fail on non-FF push

pchiusano committed:
wip

aryairani committed:
move Maybe outward in `{lca,before}Impl`

pchiusano committed:
Plumbing code to hook up usage of codebase `before` calculation

aryairani commented on parser-typechecker/src/Unison/CommandLine/InputPatterns.hs:
```diff
@@ -816,23 +818,23 @@ pushExhaustive = InputPattern
   (P.lines
     [ P.wrap $
       "The " <> makeExample' pushExhaustive <> "command can be used in place of"
-        <> makeExample' push <> "to repair remote namespaces"
+        <> makeExample' pushExhaustive <> "to repair remote namespaces"
```
oops, used "in place of `push`"

aryairani commented on parser-typechecker/src/Unison/CommandLine/InputPatterns.hs:
```diff
@@ -816,23 +818,23 @@ pushExhaustive = InputPattern
         [path] -> first fromString $ Path.parsePath' path
-        _ -> Left (I.help push)
-      Right $ Input.PushRemoteBranchI (Just (repo, path)) p SyncMode.Complete
+        _ -> Left (I.help pushExhaustive)
```
This change seems right, though.

aryairani commented on parser-typechecker/src/Unison/Codebase/FileCodebase.hs:
```diff
@@ -292,13 +293,16 @@ pushGitRootBranch
+pushGitRootBranch syncToDirectory cache onEmptyDest branch repo syncMode = do
   -- Pull the remote repo into a staging directory
   (remoteRoot, remotePath) <- viewRemoteBranch' cache (repo, Nothing, Path.empty)
+  when (remoteRoot == Branch.empty && onEmptyDest == AbortOnEmptyDest) $
+    throwError $ GitError.PushDestinationIsEmpty repo
```
This is checking if the repo root is empty, but I think we instead want it to check the target sub-namespace before we get in here.

aryairani committed:
skeleton of reformed SqliteCodebase.pushGitRootBranch

aryairani committed:
more

create v2 schema in empty repos
but don't require rootBranch
also use savepoint/release instead of transactions

aryairani committed:
remove unfriendly semicolon in ancestorsSql

unfriendly to composition; apparently not being tested anywhere yet

aryairani committed:
short-circuit ancestor search in SqliteCodebase.before implementation

also rename parts of ancestorSql to help with composability

aryairani committed:
add a test that non-fast-forward pushes produce an error,

though as of now it does not produce an error, for reasons unknown

aryairani committed:
fix white space in before query

aryairani committed:
forgot to use a `ucm` block in my transcripts

pchiusano commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs:
```diff
@@ -960,6 +966,16 @@ syncProgress = Sync.Progress need done warn allDone
       where
         v = const ()
 
+-- This might be handy later, if only as a starting point for something that ultimately returns a Connection, else delete
+-- getCodebaseForRepo :: forall m. MonadIO m => RemoteRepo -> m (Either GitError (m (), Codebase m Symbol Ann))
+-- getCodebaseForRepo repo = runExceptT @GitError @m do
```
Maybe delete this?

aryairani replied:
deleted

pchiusano commented on parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs:
```diff
@@ -1005,47 +1021,91 @@ viewRemoteBranch' (repo, sbh, path) = runExceptT do
+      chId <- fromMaybe err <$> Q.loadCausalHashIdByCausalHash h2
+      Q.setNamespaceRoot chId
+
+    -- stageAndPush remotePath = do
+    --   let repoString = Text.unpack $ printRepo repo
+    --   withStatus ("Staging codebase for upload to " ++ repoString ++ " ...") do
+    --     time "SqliteCodebase.pushGitRootBranch.stageAndPush.syncToDirectory" $ lift (syncToDirectory remotePath syncMode branch)
+    --     setRepoRoot remotePath (Branch.headHash branch)
+    --   -- push staging area to remote
+    --   withStatus ("Uploading to " ++ repoString ++ " ...") $
+    --     unlessM
+    --       ( push remotePath repo
+    --           `withIOError` (throwError . GitError.PushException repo . show)
+    --       )
+    --       (throwError $ GitError.PushNoOp repo)
```
Maybe delete

pchiusano reviewed (commented):
Left a couple comments

aryairani committed:
delete Before Command aka Command.Before

but you can revert this commit later if you want it!

aryairani committed:
delete some leftover comments
